### ISSUE TITLE
out of memory error

### ISSUE THREAD

**Issue by tomersagi** [ldbc/ldbc_snb_datagen_spark#4]
Hi,
Trying to run SF30 on 24 threads ran into the following issue:

```
15/02/11 00:24:21 WARN mapred.LocalJobRunner: job_local940749219_0013
java.lang.Exception: java.lang.OutOfMemoryError: Java heap space
at org.apache.hadoop.mapred.LocalJobRunner$Job.runTasks(LocalJobRunner.java:462)
at org.apache.hadoop.mapred.LocalJobRunner$Job.run(LocalJobRunner.java:529)
Caused by: java.lang.OutOfMemoryError: Java heap space
at ldbc.snb.datagen.objects.Knows.<init>(Knows.java:66)
at ldbc.snb.datagen.generator.KnowsGenerator.createKnow(KnowsGenerator.java:61)
at ldbc.snb.datagen.generator.KnowsGenerator.generateKnows(KnowsGenerator.java:35)
at ldbc.snb.datagen.hadoop.HadoopKnowsGenerator$HadoopKnowsGeneratorReducer.reduce(HadoopKnowsGenerator.java:43)
at ldbc.snb.datagen.hadoop.HadoopKnowsGenerator$HadoopKnowsGeneratorReducer.reduce(HadoopKnowsGenerator.java:25)
at org.apache.hadoop.mapreduce.Reducer.run(Reducer.java:171)
at org.apache.hadoop.mapred.ReduceTask.runNewReducer(ReduceTask.java:627)
at org.apache.hadoop.mapred.ReduceTask.run(ReduceTask.java:389)
at org.apache.hadoop.mapred.LocalJobRunner$Job$ReduceTaskRunnable.run(LocalJobRunner.java:319)
at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:471)
at java.util.concurrent.FutureTask.run(FutureTask.java:262)
at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1145)
at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:615)
at java.lang.Thread.run(Thread.java:745)
```

Configuration ideas? Should I increase the general heap space / site heap space / job heap space?
BTW, the machine itself has 378GB RAM. 
Thanks
Tomer

**Comment by ArnauPrat** [comment]
Hi @tomersagi 
This is probably a hadoop configuration issue, as I generate SF30 easily in a 32GB machine. Try to look at how much memory you assign to your mapper/reducer containers, as well as how much java memory you assign to mappers/reducers. For example, in my $HADOOP/etc/hadoop/mapred-site.xml I have:
```
    <property>
        <name>mapreduce.map.memory.mb</name>
        <value>4096</value>
    </property>
    <property>
        <name>mapreduce.reduce.memory.mb</name>
        <value>8192</value>
    </property>
    <property>
        <name>mapreduce.map.java.opts</name>
        <value>-Xmx3072m</value>
    </property>
    <property>
        <name>mapreduce.reduce.java.opts</name>
        <value>-Xmx6144m</value>
    </property>
```
which sets the amount of memory per container and then per mapper/reducer. Also, in your $HOME/etc/hadoop/yarn-site.xml, I have:
```    
<property>
        <name>yarn.nodemanager.resource.memory-mb</name>
        <value>100000</value>
    </property>
    <property>
        <name>yarn.scheduler.minimum-allocation-mb</name>
        <value>2048</value>
    </property>
```
Which is used to set the minimum chunk of memory allocated by hadoop and the total amount of memory it can consume. Maybe you can take a look here:
http://www.alexjf.net/blog/distributed-systems/hadoop-yarn-installation-definitive-guide/

Regards,
Arnau

**Comment by tomersagi** [comment]
Thanks, perhaps update the configuration documentation to suggest this. 
In any case, now I have a GC overhead exception:

    15/02/12 00:19:14 INFO mapred.LocalJobRunner: Finishing task: attempt_local248198299_0013_r_000023_0
    15/02/12 00:19:14 INFO mapred.LocalJobRunner: reduce task executor complete.
    15/02/12 00:19:14 WARN mapred.LocalJobRunner: job_local248198299_0013
    java.lang.Exception: java.lang.OutOfMemoryError: GC overhead limit exceeded
        at org.apache.hadoop.mapred.LocalJobRunner$Job.runTasks(LocalJobRunner.java:462)
        at org.apache.hadoop.mapred.LocalJobRunner$Job.run(LocalJobRunner.java:529)
    Caused by: java.lang.OutOfMemoryError: GC overhead limit exceeded
        at ldbc.snb.datagen.objects.Person$PersonSummary.<init>(Person.java:31)
        at ldbc.snb.datagen.objects.Knows.<init>(Knows.java:65)
        at ldbc.snb.datagen.generator.KnowsGenerator.createKnow(KnowsGenerator.java:60)
        at ldbc.snb.datagen.generator.KnowsGenerator.generateKnows(KnowsGenerator.java:35)
        at ldbc.snb.datagen.hadoop.HadoopKnowsGenerator$HadoopKnowsGeneratorReducer.reduce(HadoopKnowsGenerator.java:43)
        at ldbc.snb.datagen.hadoop.HadoopKnowsGenerator$HadoopKnowsGeneratorReducer.reduce(HadoopKnowsGenerator.java:25)
        at org.apache.hadoop.mapreduce.Reducer.run(Reducer.java:171)
        at org.apache.hadoop.mapred.ReduceTask.runNewReducer(ReduceTask.java:627)
        at org.apache.hadoop.mapred.ReduceTask.run(ReduceTask.java:389)
        at org.apache.hadoop.mapred.LocalJobRunner$Job$ReduceTaskRunnable.run(LocalJobRunner.java:319)
        at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:471)
        at java.util.concurrent.FutureTask.run(FutureTask.java:262)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1145)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:615)
        at java.lang.Thread.run(Thread.java:745)

Ideas?

**Comment by ArnauPrat** [comment]
Hi Tomer,
Can I have the full console output? Although this is not the good solution, try to increase your java heap space a little bit more.
Regards,

Arnau

**Comment by tomersagi** [comment]
I tried increasing the max heap size parameter in etc/hadoop/hadoop-env.sh to 32GB and that didn’t help.
Did you mean increasing there or some place else?

Attached is the log.

BTW, are the pre-generated data files you shared with the graph analytics mailing list full graphs that can be used for the snb workload as well?
If so I can download them and save you the trouble of helping me debug the generator.

**Comment by ArnauPrat** [comment]
Hi @tomersagi 
In principle you have to increase the amount of java memory assigned to mappers/reducers. This is done at mapred-site.xml file by setting the following option (note that different options apply to reducers and mappers):
```
<property>
        <name>mapreduce.{map|reduce}.java.opts</name>
        <value>-Xmx768m</value>
        <description>Heap-size for child jvms of {maps|reducers}.</description>
    </property>
```
The value you set must be lower than the memory assigned to mappers/reducers (now I am talking about normal memory, not java). This is set at mapred-site.xml with the following option (again, different options for mappers and reducers):
```
 <property>
        <name>mapreduce.{map|reduce}.memory.mb</name>
        <value>1024</value>
        <description>Larger resource limit for {maps|reducers}.</description>
    </property>
```
Finally, as far as I know each container will execute one mapper and one reducer, therefore, you must assing enough memory to each container to be able to hold both mapper and reducer. This is done at yarn-site.xml by means of the following option:

```
 <property>
        <name>yarn.nodemanager.resource.memory-mb</name>
        <value>4096</value>
        <description>Physical memory, in MB, to be made available to running containers</description>
    </property>
```
Also, keep in mind that if you execute 24 threads, this will create 24 containers, therefore if you assign 10GB per container, this will make a total of 240GB. So maybe it is a better idea to run with less threads. In the end, a SF30 can be generated in less than an hour using 4 threads.

About available datasets, maybe you can speak with Mihai, as I think that they were hosting some of them at TUDelft servers. In case they are not available, maybe we can do something to make them available to you.

Regards,
Arnau.

**Comment by tomersagi** [comment]
Hi,
We intend on generating really big graphs so we really need to see how we can utilize our hardware to the max. This is my current configuration:

*mapred-site.xml:

    -bash-4.2$ cat /local/hadoop/etc/hadoop/mapred-site.xml
    <configuration>
    <property>
            <name>mapreduce.map.memory.mb</name>
            <value>2048</value>
    </property>
    <property>
            <name>mapreduce.reduce.memory.mb</name>
            <value>2048</value>
    </property>
    <property>
            <name>mapreduce.map.java.opts</name>
            <value>-Xmx1768m -XX:-UseGCOverheadLimit</value>
    </property>
    <property>
        <name>mapreduce.reduce.java.opts</name>
        <value>-Xmx1768m -XX:-UseGCOverheadLimit</value>
    </property>
    </configuration>

*yarn-site.xml

    <configuration>
    <property>
        <name>yarn.nodemanager.resource.memory-mb</name>
        <value>4128</value>
    </property>
    <property>
        <name>yarn.scheduler.minimum-allocation-mb</name>
        <value>2048</value>
    </property>
    </configuration>

Since I'm running 24 threads. I should be fine (24X4128=99072=~97GB and I have 378GB on this machine). I'm now not getting the GC error but rather the same error from the beginning of the issue. 
Your assistance is highly appreciated. 
Tomer

**Comment by ArnauPrat** [comment]
Hi @tomersagi 
This is my configuration for a 8 core machine with 128GB of memory (this is the same I posted above):
```
    <property>
        <name>mapreduce.map.memory.mb</name>
        <value>4096</value>
    </property>
    <property>
        <name>mapreduce.reduce.memory.mb</name>
        <value>8192</value>
    </property>
    <property>
        <name>mapreduce.map.java.opts</name>
        <value>-Xmx3072m</value>
    </property>
    <property>
        <name>mapreduce.reduce.java.opts</name>
        <value>-Xmx6144m</value>
    </property>
```
Here, you see that I am assigning 4Gb per mapper and 8GB per reducer (total memory). Out of these, 3072 and 6144 are assigned to java heap, respectively. Similarly, in my yarn-site.xml, I have the following (besides other options):
```
 <property>
        <name>yarn.nodemanager.resource.memory-mb</name>
        <value>100000</value>
    </property>
    <property>
        <name>yarn.scheduler.minimum-allocation-mb</name>
        <value>2048</value>
    </property>
```
The first sets the maximum amount of memory that can be assigned to all containers. A container is an abstraction of the underlying operating system, some sort of virtual machine with a bunch of resurces assigned (memory, cores, etc..). This helps applications to be isolated. See here for more details:
http://www.kdnuggets.com/2014/07/containers-yarn-enabler.html and http://stackoverflow.com/questions/14365218/what-is-a-container-in-yarn. Is in these containes where hadoop jobs are executed. The memory assigned to all containers cannot be greater than this parameter. Therefore, assuming a worst case scenario, where I have 8 mappers and 8 reducers running in 8 containers (one mapper and one reducer per container), Each container will need at least mapreduce.map.memory.mb + mapreduce.reduce.memory.mb = 4096 + 8192 = 12288 GB, multiplied by 8 containers = 12288*8 = 98304 mb. 98304 < 100000, therefore with this setting I should be fine. In your case, with 24 threads, you would require 12288*24 = 294912 mb at least. Note that I have not experimented a lot with these parameters, therefore smaller values of mapreduce.map.memory.mb and mapreduce.reduce.memory.mb could perfectly work as well.

The second parameter, as far as I know, is the minimum allocation bunch of data. So everytime a container asks for more memory to yarn, this is the minimum amount he can ask. I do not remember why I set this to 2048, but this will probably have an effect to the performance, specially for high memory demanding applications that continuously ask for more memory. I think that setting this value higher will reduce overheads, but of course can make you waste more memory. There is an equivalent parameter for setting the maximum block size to allocate, but I have it to default.

So, to sum up, in your case I would increase yarn.nodemanager.resource.memory-mb from 4128 to 24*4096 = 98304 ~ 100000 and see what happens. If we still have problems with memory we can try to increase the memory assigned to mappers/reducers. Also, keep in mind that in DATAGEN, the most memory consuming tasks are reducers, so it might be preferable to assign more memory to reducers than mappers, in case we were in a situation where we had limited resources and had to choose.

I hope this helps, so let's cross fingers :+1: 

Regards and thx for your patience :) with all of these issues.
Arnau

**Comment by tomersagi** [comment]
No go. 
Now I'm back to the GC overhead error. Here is the current config:

    -bash-4.2$ cat /local/hadoop/etc/hadoop/yarn-site.xml
    <?xml version="1.0"?>
    <configuration>
    <property>
        <name>yarn.nodemanager.resource.memory-mb</name>
        <value>100000</value>
    </property>
    <property>
        <name>yarn.scheduler.minimum-allocation-mb</name>
        <value>2048</value>
    </property>

    </configuration>
    -bash-4.2$ cat /local/hadoop/etc/hadoop/mapred-site.xml
    <configuration>
    <property>
        <name>mapreduce.map.memory.mb</name>
        <value>1024</value>
    </property>
    <property>
        <name>mapreduce.reduce.memory.mb</name>
        <value>3072</value>
    </property>
    <property>
        <name>mapreduce.map.java.opts</name>
        <value>-Xmx976m -XX:-UseGCOverheadLimit</value>
    </property>
    <property>
        <name>mapreduce.reduce.java.opts</name>
        <value>-Xmx2874m -XX:-UseGCOverheadLimit</value>
    </property>
    </configuration>

I'll try mimicking your config exactly and see if that works and then move on from there.

**Comment by tomersagi** [comment]
using the exact same configuration as you did with SF30, 8 threads, I got a GC overhead limit exceeded error again.

**Comment by ArnauPrat** [comment]
Hi @tomersagi

Still crashing during knows generation?

2015-02-23 10:44 GMT+01:00 Tomer Sagi <notifications@github.com>:

**Comment by tomersagi** [comment]
Resolved. I missed a step in the initial pseudo-distributed cluster configuration causing yarn to crash. 
All is well now.

**Comment by MohamedRagabAnas** [comment]
I have the same problem with generation the data of SF=30, I am  facing the same problem with GC  overhead limit exceeded !

* mapred-site.xml
```
    <property>
        <name>mapreduce.map.memory.mb</name>
        <value>8192</value>
    </property>

    <property>
        <name>mapreduce.reduce.memory.mb</name>
        <value>12288</value>
    </property>

    <property>
        <name>mapreduce.map.java.opts</name>
        <value>-Xmx6144m -XX:-UseGCOverheadLimit</value>
    </property>

    <property>
        <name>mapreduce.reduce.java.opts</name>
        <value>-Xmx9216?m -XX:-UseGCOverheadLimit</value>
    </property>

   <property>
   <name>mapreduce.reduce.shuffle.memory.limit.percent</name>
  <value>0.06</value>
```

* yarn-site.xml

```
    <property>
        <name>yarn.nodemanager.resource.memory-mb</name>
        <value>100000</value>
    </property>

    <property>
        <name>yarn.scheduler.minimum-allocation-mb</name>
        <value>2048</value>
    </property>
```

I tried with the same values of @tomersagi , but I think that I have only 1 thread running .

@tomersagi  could you please tell me, how could you solved the GC error!
@ArnauPrat  your help will be totally appreciated !
Thanks in advance for your help

**Comment by szarnyasg** [comment]
@MohamedRagabAnas the question mark here seems suspicious.
```xml
    <property>
        <name>mapreduce.reduce.java.opts</name>
        <value>-Xmx9216?m -XX:-UseGCOverheadLimit</value>
    </property>
```

What's your setup? How many nodes are in your Hadoop cluster?

**Comment by MohamedRagabAnas** [comment]
Hi @szarnyasg , sorry the question mark here was put by mistake, 
I'm using a single machine centos with 256GB Ram .
Thanks in advance Gabour for your help

**Comment by szarnyasg** [comment]
On a single machine, I usually use the `HADOOP_CLIENT_OPTS` environment variable. Before running the script, issue the following command:
```bash
export HADOOP_CLIENT_OPTS="-Xmx200G"
```
Hope this helps.

**Comment by MohamedRagabAnas** [comment]
Thanks Gabour, this works for me :)
